File: _scripts/update-ids.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Script to manually update spell IDs in keybinds.yml
# Usage: ruby _scripts/update-ids.rb [class] [key] [spell_id]
# Example: ruby _scripts/update-ids.rb warrior q 23922

require 'yaml'

KEYBINDS_FILE = '_data/keybinds.yml'

def load_keybinds
  YAML.load_file(KEYBINDS_FILE)
rescue Errno::ENOENT
  puts "Error: #{KEYBINDS_FILE} not found"
  exit 1
end

def save_keybinds(keybinds_data)
  File.write(KEYBINDS_FILE, keybinds_data.to_yaml)
end

def update_id(class_key, key_id, spell_id)
  keybinds_data = load_keybinds
  
  unless keybinds_data['classes'][class_key]
    puts "Error: Class '#{class_key}' not found"
    puts "Available classes: #{keybinds_data['classes'].keys.join(', ')}"
    return false
  end
  
  unless keybinds_data['classes'][class_key]['abilities'][key_id]
    puts "Error: Key '#{key_id}' not found for class '#{class_key}'"
    return false
  end
  
  old_id = keybinds_data['classes'][class_key]['abilities'][key_id]['wowhead_id']
  keybinds_data['classes'][class_key]['abilities'][key_id]['wowhead_id'] = spell_id.to_i
  
  ability_name = keybinds_data['classes'][class_key]['abilities'][key_id]['name']
  class_name = keybinds_data['classes'][class_key]['name']
  
  puts "Updated: #{class_name} - #{key_id.upcase} (#{ability_name})"
  puts "  Old ID: #{old_id || 'null'}"
  puts "  New ID: #{spell_id}"
  
  save_keybinds(keybinds_data)
  puts "Changes saved to #{KEYBINDS_FILE}"
  true
end

def list_missing
  keybinds_data = load_keybinds
  missing = []
  
  keybinds_data['classes'].each do |class_key, class_data|
    next unless class_data['abilities']
    
    class_data['abilities'].each do |key_id, ability|
      if ability.is_a?(Hash) && (ability['wowhead_id'].nil? || ability['wowhead_id'] == 'null')
        missing << {
          'class' => class_key,
          'class_name' => class_data['name'],
          'key' => key_id,
          'ability_name' => ability['name']
        }
      end
    end
  end
  
  if missing.empty?
    puts "All abilities have spell IDs!"
  else
    puts "Missing spell IDs (#{missing.length} total):\n\n"
    missing.each do |item|
      puts "#{item['class_name']} - #{item['key'].upcase}: #{item['ability_name']}"
    end
  end
end

def main
  if ARGV.empty? || ARGV[0] == '--list' || ARGV[0] == '-l'
    list_missing
  elsif ARGV.length == 3
    class_key = ARGV[0]
    key_id = ARGV[1].downcase
    spell_id = ARGV[2]
    
    update_id(class_key, key_id, spell_id)
  else
    puts "Usage:"
    puts "  ruby _scripts/update-ids.rb [class] [key] [spell_id]"
    puts "  ruby _scripts/update-ids.rb --list  (list all missing IDs)"
    puts ""
    puts "Example:"
    puts "  ruby _scripts/update-ids.rb warrior q 23922"
    puts ""
    puts "Available classes: warrior, paladin, deathknight, demonhunter, druid, monk"
  end
end

main if __FILE__ == $PROGRAM_NAME

File: scripts/update-ids.ts
#!/usr/bin/env node
// Ports _scripts/update-ids.rb — manually update a spell ID in keybinds.yml.
// Usage: node scripts/update-ids.ts [class] [key] [spell_id]
// Example: node scripts/update-ids.ts warrior q 23922

import fs from 'node:fs';
import { parseKeybinds, serializeKeybinds } from '../src/lib/keybinds-io.ts';
import { KEYBINDS_PATH } from './paths.ts';
import type { KeybindsData } from '../src/lib/types.ts';

function loadKeybinds(): KeybindsData {
	return parseKeybinds(fs.readFileSync(KEYBINDS_PATH, 'utf8'));
}

function saveKeybinds(data: KeybindsData): void {
	fs.writeFileSync(KEYBINDS_PATH, serializeKeybinds(data), 'utf8');
}

function updateId(classKey: string, keyId: string, spellId: string): boolean {
	const keybinds = loadKeybinds();
	const classData = keybinds.classes[classKey];

	if (!classData) {
		console.log(`Error: Class '${classKey}' not found`);
		console.log(`Available classes: ${Object.keys(keybinds.classes).join(', ')}`);
		return false;
	}

	const ability = classData.abilities[keyId];
	if (!ability) {
		console.log(`Error: Key '${keyId}' not found for class '${classKey}'`);
		return false;
	}

	const oldId = ability.wowhead_id;
	ability.wowhead_id = Number(spellId);

	console.log(`Updated: ${classData.name} - ${keyId.toUpperCase()} (${ability.name})`);
	console.log(`  Old ID: ${oldId ?? 'null'}`);
	console.log(`  New ID: ${spellId}`);

	saveKeybinds(keybinds);
	console.log(`Changes saved to ${KEYBINDS_PATH}`);
	return true;
}

function listMissing(): void {
	const keybinds = loadKeybinds();
	const missing: { className: string; key: string; abilityName: string }[] = [];

	for (const classData of Object.values(keybinds.classes)) {
		for (const [keyId, ability] of Object.entries(classData.abilities ?? {})) {
			const id = ability?.wowhead_id;
			if (id === undefined || id === null || String(id) === 'null') {
				missing.push({ className: classData.name, key: keyId, abilityName: ability.name });
			}
		}
	}

	if (missing.length === 0) {
		console.log('All abilities have spell IDs!');
	} else {
		console.log(`Missing spell IDs (${missing.length} total):\n`);
		for (const item of missing) {
			console.log(`${item.className} - ${item.key.toUpperCase()}: ${item.abilityName}`);
		}
	}
}

function main(): void {
	const args = process.argv.slice(2);

	if (args.length === 0 || args[0] === '--list' || args[0] === '-l') {
		listMissing();
	} else if (args.length === 3) {
		const [classKey, keyArg, spellId] = args;
		updateId(classKey, keyArg.toLowerCase(), spellId);
	} else {
		console.log('Usage:');
		console.log('  node scripts/update-ids.ts [class] [key] [spell_id]');
		console.log('  node scripts/update-ids.ts --list  (list all missing IDs)');
		console.log('');
		console.log('Example:');
		console.log('  node scripts/update-ids.ts warrior q 23922');
		console.log('');
		console.log('Available classes: warrior, paladin, deathknight, demonhunter, druid, monk');
	}
}

main();
